ocaml/testsuite/tests/typing-layouts-err-msg/value.ml

```diff
@@ -0,0 +1,433 @@
+|}]
+
+(* Arrow *)
+type ('a: void) t = 'a
+let f (x : int -> int): 'a t = x
+[%%expect{|
+type ('a : void) t = 'a
+Line 2, characters 31-32:
+2 | let f (x : int -> int): 'a t = x
+                                   ^
+Error: This expression has type int -> int
+       but an expression was expected of type 'a t = ('a : void)
+       The layout of int -> int is value, because
+         it's a function type.
+       But the layout of int -> int must be a sublayout of void, because
+         of the definition of t at line 1, characters 0-22.
+|}]
+
+(* Tfield *)
+(* CR layouts v2.9: add test *)
+
+(* Tnil *)
+(* CR layouts v2.9: add test *)
+
+(* First_class_module *)
+type ('a: void) t = 'a
+module type X_int = sig val x : int end;;
+module Three : X_int = struct let x = 3 end;;
+let f (): 'a t = (module Three : X_int)
+[%%expect{|
+type ('a : void) t = 'a
+module type X_int = sig val x : int end
+module Three : X_int
+Line 4, characters 17-39:
+4 | let f (): 'a t = (module Three : X_int)
+                     ^^^^^^^^^^^^^^^^^^^^^^
+Error: This expression has type (module X_int)
+       but an expression was expected of type 'a t = ('a : void)
+       The layout of (module X_int) is value, because
+         it's a first-class module type.
+       But the layout of (module X_int) must be a sublayout of void, because
+         of the definition of t at line 1, characters 0-22.
+|}]
+
+(* Separability_check *)
+(* Only used within [Result.is_error] and not thrown as an exception *)
+
+(* Univar *)
+let f: 'a. 'a -> ('b : void) = fun x -> x
+[%%expect{|
+Line 1, characters 40-41:
+1 | let f: 'a. 'a -> ('b : void) = fun x -> x
+                                            ^
+Error: This expression has type ('a : value)
+       but an expression was expected of type ('b : void)
+       The layout of 'b is void, because
+         of the annotation on the type variable 'b.
+       But the layout of 'b must overlap with value, because
+         it's an unannotated universal variable.
+|}];;
+
+(* Polymorphic_variant_field *)
+let f (x : t_float64) = `A x;;
+[%%expect{|
+Line 1, characters 27-28:
+1 | let f (x : t_float64) = `A x;;
+                               ^
+Error: This expression has type t_float64
+       but an expression was expected of type ('a : value)
+       The layout of t_float64 is float64, because
+         of the definition of t_float64 at line 5, characters 0-24.
+       But the layout of t_float64 must be a sublayout of value, because
+         it's a field of a polymorphic variant.
+|}];;
+
+(* Default_type_layout *)
+type t
+let f : t -> ('a : void) = fun x -> x
+[%%expect{|
+type t
+Line 2, characters 36-37:
+2 | let f : t -> ('a : void) = fun x -> x
+                                        ^
+Error: This expression has type t but an expression was expected of type
+         ('a : void)
+       The layout of t is value, because
+         of the definition of t at line 1, characters 0-6.
```

Comment:
This isn't the message we're looking for. I think you'll need a mutually recursive group of type declarations.